### PR TITLE
Introduce apictl gen deployment-dir command to generate deployment directory using the source

### PR DESCRIPTION
**Pull request by Chamindu36**
## Purpose
Introduce a new command as apictl generate. This will generate the Deployment folder under a given mode. In K8 mode, the new generate command will bring values in the param file into a set of config maps.

## Goals
Fixes https://github.com/wso2/product-apim-tooling/issues/555

## Approach

The newly introduced commands can be in the below form.

#### **Command**

> _**apictl gen deployment-dir [flags]**_

#### Examples

`apictl    gen deployment-dir    -s  /chamindu/desktop/source/Dev/PizzaShackAPI_1.0.0_r1.zip    -d   /chamindu/desktop/deployment/Dev`

##### Mandatory

-s, --source | string | Filepath of the source artifact which is needed to use when generating deployment directory
-- | -- | --

##### Optional

-d, --destination | string | Path of the directory where the new deployment directory should be generated
-- | -- | --

When generating this directory structure we have to consider both the VM use case and K8S operator use case. The functionalities and folder structures will be different from each other based on the use case. 

### VM Mode
The folder structure to follow when generating a deployment directory for a particular API will look like this.

```

└── <Name_of_the_deployment_directoy>
    └── api_params.yaml
    └── api_meta.yaml (This file will contain the details related to artefact)
    └── certificates (This will be a directory to include certificates)

```

## Command Overview

Screenshot from 2020-12-18 14-41-42

Screenshot from 2020-12-18 14-42-06

## Documentation
Documentation changes need to be done.

### CODE REVIEW

File: import-export-cli/cmd/gen.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+package cmd
+
+import (
+	"github.com/spf13/cobra"
+	"github.com/wso2/product-apim-tooling/import-export-cli/utils"
+)
+
+// Get command related usage Info
+const GenCmdLiteral = "gen"
+const GenCmdShortDesc = "Generate deployment directory for VM and for K8S operator"
```
Comment (npamudika) [inline import-export-cli/cmd/gen.go:28]:
```suggestion
const GenCmdShortDesc = "Generate deployment directory for VM and K8S operator"
```

File: import-export-cli/cmd/gen.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+const GenCmdLongDesc = `Generate sample directory with all the contents to use as the deployment directory` +
+	`  when performing CI/CD pipeline `
```
Comment (npamudika) [inline import-export-cli/cmd/gen.go:31]:
```suggestion
	`  when performing CI/CD pipeline tasks`
```

File: import-export-cli/cmd/genDeploymentDir.go [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+	"path/filepath"
+	"strings"
+)
+
+var genDeploymentDirDestination string
+var genDeploymentDirSource string
+
+// GetEnvsCmd related info
+const GenDeploymentDirCmdLiteral = "deployment-dir"
+const GenDeploymentDirCmdShortDesc = "Generate sample deployment directory"
```
Comment (npamudika) [inline import-export-cli/cmd/genDeploymentDir.go:37]:
```suggestion
const GenDeploymentDirCmdShortDesc = "Generate a sample deployment directory"
```

File: import-export-cli/cmd/genDeploymentDir.go [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+
+const GenDeploymentDirCmdLongDesc = `Generate sample deployment directory based on the provided source artifact`
```
Comment (npamudika) [inline import-export-cli/cmd/genDeploymentDir.go:39]:
```suggestion
const GenDeploymentDirCmdLongDesc = `Generate a sample deployment directory based on the provided source artifact`
```

File: import-export-cli/cmd/genDeploymentDir.go [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+		pwd, err := os.Getwd()
+		if err != nil {
+			return err
+		}
+		deploymentDirParent = pwd
+	}
+
+	// Check whether the source is existed in the given location
+	if _, err := os.Stat(genDeploymentDirSource); os.IsNotExist(err) {
+		utils.HandleErrorAndContinue("Error retrieving the source file from the given path ", err)
```
Comment (npamudika) [inline import-export-cli/cmd/genDeploymentDir.go:89]:
```suggestion
		utils.HandleErrorAndContinue("Error retrieving the source file from the given path " + sourceDirectoryPath + " ", err)
```

File: import-export-cli/cmd/genDeploymentDir.go [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+				if !stat.IsDir() {
+					fmt.Printf("%s is not a directory\n", genDeploymentDirDestination)
+					os.Exit(1)
+				}
+			}
+		}
+
+		err := executeGenDeploymentDirCmd()
+		if err != nil {
+			utils.HandleErrorAndContinue("Error initializing Deployment directory", err)
```
Comment (npamudika) [inline import-export-cli/cmd/genDeploymentDir.go:200]:
```suggestion
			utils.HandleErrorAndContinue("Error initializing the Deployment directory", err)
```

File: import-export-cli/cmd/genDeploymentDir.go [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+		}
+	},
+}
+
+func init() {
+	GenCmd.AddCommand(genDeploymentDirCmd)
+	genDeploymentDirCmd.Flags().StringVarP(&genDeploymentDirDestination, "destination", "d", "", "Path of "+
+		"the directory where the directory should be generated")
+	genDeploymentDirCmd.Flags().StringVarP(&genDeploymentDirSource, "source", "s", "", "Path of "+
+		"the source directory to use when generating the directory should be generated")
```
Comment (npamudika) [inline import-export-cli/cmd/genDeploymentDir.go:210]:
```suggestion
		"the source directory to be used when generating the directory")
```

File: import-export-cli/cmd/gen.go [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+package cmd
+
+import (
+	"github.com/spf13/cobra"
+	"github.com/wso2/product-apim-tooling/import-export-cli/utils"
+)
+
+// Get command related usage Info
+const GenCmdLiteral = "gen"
```
Comment (wasuradananjith) [inline import-export-cli/cmd/gen.go:27]: Shall we run "go run tools/gen.go" to generate the docs (.md files) to the whole feature and commit those files here?
Also, shall we add the commands and the usage of this feature to the file in [1] as well?

[1] https://github.com/wso2/product-apim-tooling/blob/master/import-export-cli/resources/README.html

Comment (Chamindu36) [reply import-export-cli/cmd/gen.go:27]: Added the genarated documents and added the new command details to README.html in [1]. Thanks for pointing this out.
[1] https://github.com/wso2/product-apim-tooling/blob/master/import-export-cli/resources/README.html

**Review by npamudika — APPROVED** [review]

**Review by wasuradananjith — APPROVED** [review]

File: import-export-cli/impl/common.go [hunk 9]
```diff
@@ -134,3 +137,35 @@ func WriteTargetFileFromTemplate(targetFile string, tmpl []byte, envs *utils.Mai
...
+	if err != nil {
+		utils.HandleErrorAndExit("Error creating api_meta.yaml inside the exported zip archive", err)
+	}
+
+	err = utils.Zip(tmpClonedLoc, targetZipFile)
+	if err != nil {
+		utils.HandleErrorAndExit("Error creating the final zip archive", err)
+	}
+	return nil
+}
```
Comment (menakaj) [inline import-export-cli/impl/common.go:171]: Add new line

File: import-export-cli/box/resources/sample/api_params.yaml [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+          - hostName:
+            alias:
+            path:
+        mutualSslCerts:
+          - tierName:
+            alias:
+            path:
+          - tierName:
+            alias:
+            path:
```
Comment (Meruja) [inline import-export-cli/box/resources/sample/api_params.yaml:20]: Shall we add a line at the eof?

Comment (Chamindu36) [reply import-export-cli/box/resources/sample/api_params.yaml:20]: Fixed with f75a26712774c479a96a139dc5fdc95d089bb104

File: import-export-cli/impl/common.go (hunk 9 above)
Comment (Chamindu36) [reply import-export-cli/impl/common.go:171]: Fixed with f75a26712774c479a96a139dc5fdc95d089bb104

**Pull request merged**